src/ec/system76/ec/system76_ec.c

```diff
@@ -118,33 +123,529 @@ uint8_t system76_ec_read_version(uint8_t *data)
+
+	uint8_t read_cmd[2] = {
+		CMD_SPI_FLAG_READ | CMD_SPI_FLAG_SCRATCH,
+		0,
+	};
+
+	for (addr = 0; addr + SPI_ACCESS_SIZE < len; addr += SPI_ACCESS_SIZE) {
+		read_cmd[1] = SPI_ACCESS_SIZE;
+
+		if ((rv = system76_ec_smfi_cmd(CMD_SPI, ARRAY_SIZE(read_cmd), read_cmd)))
+			return rv;
+
+		if (system76_ec_read(REG_DATA + 1) != read_cmd[1])
+			return 1;
+
+		for (i = 0; i < read_cmd[1]; i++)
+			dest[addr + i] = system76_ec_read(REG_DATA + 2 + i);
+	}
+
+	/* All chunks read? */
+	if (addr == len)
+		return 0;
+
+	/* Read last chunk */
+	read_cmd[1] = len % SPI_ACCESS_SIZE;
+
+	if ((rv = system76_ec_smfi_cmd(CMD_SPI, ARRAY_SIZE(read_cmd), read_cmd)))
+		return rv;
+
+	if (system76_ec_read(REG_DATA + 1) != read_cmd[1])
+		return 1;
+
+	for (i = 0; i < read_cmd[1]; i++)
+		dest[addr + i] = system76_ec_read(REG_DATA + 2 + i);
+
+	return 0;
+}
+
+/*
+ * Write len bytes from data to EC SPI bus
+ * Returns 0 on success, >0 on error
+ */
+static uint32_t ec_spi_bus_write(uint8_t *data, uint8_t len)
+{
+	uint32_t addr, i, rv;
+	uint8_t write_cmd[2] = {
+		[0] = CMD_SPI_FLAG_SCRATCH,
+		[1] = len,
+	};
+
+	for (addr = 0; addr + SPI_ACCESS_SIZE < len; addr += SPI_ACCESS_SIZE) {
+		write_cmd[1] = SPI_ACCESS_SIZE;
+
+		for (i = 0; i < write_cmd[1]; ++i)
+			system76_ec_write(REG_DATA + 2 + i, data[addr + i]);
+
+		if ((rv = system76_ec_smfi_cmd(CMD_SPI, ARRAY_SIZE(write_cmd), write_cmd)))
+			return rv;
+
+		if (system76_ec_read(REG_DATA + 1) != write_cmd[1])
+			return 1;
+	}
+
+	/* All chunks written? */
+	if (addr == len)
+		return 0;
+
+	/* Write last chunk */
+	write_cmd[1] = len % SPI_ACCESS_SIZE;
+
+	for (i = 0; i < write_cmd[1]; i++)
+		system76_ec_write(REG_DATA + 2 + i, data[addr + i]);
+
+	if ((rv = system76_ec_smfi_cmd(CMD_SPI, ARRAY_SIZE(write_cmd), write_cmd)))
+		return rv;
+
+	if (system76_ec_read(REG_DATA + 1) != write_cmd[1])
+		return 1;
+
+	return 0;
+}
+
+static int ec_spi_cmd_status(void)
+{
+	uint8_t rv;
+	uint8_t buf[1];
+
+	if ((rv = ec_spi_reset()))
+		return -rv;
+
+	buf[0] = 0x05; /* SPI Read Status */
+
+	if ((rv = ec_spi_bus_write(buf, 1)))
+		return -rv;
+
+	if ((rv = ec_spi_bus_read(buf, 1)))
+		return -rv;
+
+	return buf[0];
+}
+
+static int ec_spi_cmd_write_enable(void)
+{
+	int status;
+	uint8_t rv;
+	uint8_t buf[1];
+
+	if ((rv = ec_spi_reset()))
+		return -rv;
+
+	buf[0] = 0x06; /* SPI Write Enable */
+
+	if ((rv = ec_spi_bus_write(buf, 1)))
+		return -rv;
+
+	do {
+		status = ec_spi_cmd_status();
+	} while (status > 0 && (status & 3) != 2);
+
+	return 0;
+}
+
+static int ec_spi_cmd_write_disable(void)
+{
+	int status;
+	uint8_t rv;
+	uint8_t buf[1];
+
+	if ((rv = ec_spi_reset()))
+		return -rv;
+
+	buf[0] = 0x04; /* SPI Write Disable */
+
+	if ((rv = ec_spi_bus_write(buf, 1)))
+		return -rv;
+
+	do {
+		status = ec_spi_cmd_status();
+	} while (status > 0 && (status & 3) != 0);
+
+	return 0;
+}
+
+/* Erase a sector at addr. Returns 0 on success <0 on error. */
+static long ec_spi_erase_sector(uint32_t addr)
```

Comment:
If it returns negative value then it should be `int` probably

Reply:
`long` is also signed, but yeah, `int` is enough.

Reply:
https://github.com/Dasharo/coreboot/pull/481/commits/5404a326dbe4086d0596d4b7dbd7f4c1ad32cef0